include/openmc/tallies/filter.h

```diff
@@ -76,6 +72,17 @@ class Filter
   int n_bins_;
 };
 
+//==============================================================================
+// Global variables
+//==============================================================================
+
+extern "C" int32_t n_filters;
+
+extern std::vector<FilterMatch> filter_matches;
+#pragma omp threadprivate(filter_matches)
+
+extern std::vector<Filter*> tally_filters;
```

Comment:
I kind of like having the global variables up at the top of the header files.  It means that we have to forward declare everything, but the declarations of global variables are really small (usually only 1 LoC) and can be easily missed otherwise.

Reply:
I generally agree. However, in order for the Intel compiler to accept the `threadprivate` directive, the variable type cannot be incomplete, i.e., a forward declaration is not sufficient because the (Intel) compiler needs to know the size of `FilterMatch` in order to set up thread-local storage. I also tried moving only the pragma after the declaration of `FilterMatch`, but alas the compiler was still not happy.

Reply:
Hmm, come to think of it, it may be possible to make an explicit instantiation of `std::vector<FilterMatch>` with `FilterMatch` being forward declared because I don't think `std::vector` needs to know the size of its template argument (presumably it just stores pointers to the type). Let me give that a shot...

Reply:
Looks like you're right:
```
/usr/include/c++/7.3.0/bits/stl_vector.h(1467): error: incomplete type is not allowed
  	typename aligned_storage<sizeof(_Tp), alignof(_Tp)>::type __buf;
```